### PR TITLE
Add a gradle task to check compability with child components

### PR DESCRIPTION
**Pull request by gaiksaya**
### Description
This change adds a gradle task to check the compatibility with any child projects/components. The task is dependent on `publishToMavenLocal` as it publishes the core engine to maven local and then clones the remote project in temporary directory and runs `./gradlew assemble`.
The task takes 2 inputs:
1. repositoryUrls: a comma separated git url of the components. Defaults to https://raw.githubusercontent.com/opensearch-project/opensearch-plugins/main/plugins/.meta
2. ref: branch or reference to checkout. Defaults to `main`

Run with specific components:
```
./gradlew checkCompatibility
> Task :checkCompatibility
Incompatible components: [git@github.com:opensearch-project/k-nn.git, git@github.com:opensearch-project/asynchronous-search.git, git@github.com:opensearch-project/anomaly-detection.git, git@github.com:opensearch-project/observability.git, git@github.com:opensearch-project/performance-analyzer-rca.git, git@github.com:opensearch-project/performance-analyzer.git, git@github.com:opensearch-project/reporting.git, git@github.com:opensearch-project/geospatial.git, git@github.com:opensearch-project/sql.git, git@github.com:opensearch-project/cross-cluster-replication.git, git@github.com:opensearch-project/opensearch-oci-object-storage.git, git@github.com:opensearch-project/neural-search.git]
Compatible components: [git@github.com:opensearch-project/job-scheduler.git, git@github.com:opensearch-project/common-utils.git, git@github.com:opensearch-project/notifications.git, git@github.com:opensearch-project/security.git, git@github.com:opensearch-project/alerting.git, git@github.com:opensearch-project/security-analytics.git, git@github.com:opensearch-project/index-management.git, git@github.com:opensearch-project/ml-commons.git]

BUILD SUCCESSFUL in 11m 49s
983 actionable tasks: 821 executed, 90 from cache, 72 up-to-date
```

Run with all components:

```
./gradlew checkCompatibility -PrepositoryUrls=https://github.com/opensearch-project/job-scheduler.git,https://github.com/opensearch-project/common-utils.git,https://github.com/opensearch-project/asynchronous-search.git
> Task :checkCompatibility
Incompatible components: [https://github.com/opensearch-project/asynchronous-search.git]
Compatible components: [https://github.com/opensearch-project/common-utils.git, https://github.com/opensearch-project/job-scheduler.git]

BUILD SUCCESSFUL in 1m 57s
837 actionable tasks: 259 executed, 578 up-to-date

BUILD SUCCESSFUL in 2m 23s
837 actionable tasks: 259 executed, 578 up-to-date
```
Run with specific branch
```
./gradlew checkCompatibility -PrepositoryUrls=https://github.com/opensearch-project/common-utils -Pref=lol

> Task :checkCompatibility
Skipping compatibility check for https://github.com/opensearch-project/common-utils
Components skipped due to git failures: [https://github.com/opensearch-project/common-utils]

BUILD SUCCESSFUL in 14s
837 actionable tasks: 255 executed, 582 up-to-date
```
### Related Issues
https://github.com/opensearch-project/opensearch-devops/issues/114#issuecomment-1548868449

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by reta** [comment]
@gaiksaya I understand the idea but I think we could (we should) model it as the Github worklfow, the OpenSearch core should not have any particular knowledge about the components that depends on it.

**Comment by reta** [comment]
For certain plugins, would be good to use `publishPluginZipPublicationToMavenLocal` task, not all plugins have JARs to publish

**Comment by gaiksaya** [comment]
> @gaiksaya I understand the idea but I think we could (and probably should) model it as the Github worklfow, the OpenSearch core should not have any particular knowledge about the components that depends on it.

That's right! My next step would be adding a GHA workflow to run this task. Also this task is optional which will run if and when someone wants to check the compatibility against any remote project. Maybe @nknize can help explain in better words?

>For certain plugins, would be good to use publishPluginZipPublicationToMavenLocal task, not all plugins have JARs to publish

The reason I am publishing everything to mavenLocal is because only we do not know what module the child component is using. So just publish to mavenlocal if the task exists for all sub-modules and then check the compatibility.

**Comment by reta** [comment]
> That's right! My next step would be adding a GHA workflow to run this task.

To clarify, I think we should not make any changes in OpenSearch build, the workflow could:
 - checkout and publish OpenSearch core
 - checkout and publish dependent components

> The reason I am publishing everything to mavenLocal is because only we do not know what module the child component is using.

By running `./gradlew tasks` it is easy to see which once the child component support, publishToMavenLocal or publishPluginZipPublicationToMavenLocal

**Comment by gaiksaya** [comment]
> To clarify, I think we should not make any changes in OpenSearch build

When we run directly on GHA, I believe it will be a little late for the PR author to detect what all components (mostly from among the distribution components) are they breaking. In this way, the contributor can check for compatibility even before the PR is raised. 

> By running ./gradlew tasks it is easy to see which once the child component support, publishToMavenLocal or publishPluginZipPublicationToMavenLocal

Sorry for confusion. What I meant was, instead of grabbing the dependencies of OpenSearch from https://aws.oss.sonatype.org/content/repositories/snapshots/ which is updated only when a PR is merged/pushed, publishing everything to maven local is fail fast strategy.
For example: I checked out this PR (when it was in review) https://github.com/opensearch-project/OpenSearch/pull/7646/ and added the gradle task to check compatibility with all plugins. Since the PR was not merged [snapshots repo](https://aws.oss.sonatype.org/content/repositories/snapshots/) is not updated but publishing to maven local and plugins [looking for dependencies in the mavenlocal first ](https://github.com/opensearch-project/anomaly-detection/blob/main/build.gradle#L55) fail with recent changes.
<details><summary>Failure logs because of the above PR in AD plugin</summary>
<p>

```
/tmp/groovy-generated-tmpdir-5275915656943254951/src/main/java/org/opensearch/ad/transport/AnomalyResultTransportAction.java:85: error: package org.opensearch.common.lease does not exist
import org.opensearch.common.lease.Releasable;
```

</p>
</details>

**Comment by gaiksaya** [comment]
>  On this point, some changes are intentionally breaking the APIs, leading to chicken and egg problem - the failing check will prevent the merge but without the merge it is impossible to change the dependent component.

The intention is to just notify the failing components via GH issue. Also the merge won't be blocked even if the component fails (unless the gradle task itself breaks). See PR description for sample run. The changes will merge as intended in the core repo but just inform (as heads up) to the child components. If the author of the PR wishes to fix them they can else just create the issue and move on.

**Comment by reta** [comment]
> The intention is to just notify the failing components via GH issue.

I missed that part, sorry about it (for some reasons expected check to fail)

**Comment by dblock** [comment]
I like the intent of the task, but I'm with @reta on actually implementing some parts of this as a Gradle task. I don't love that it pulls down a list of dependent components from GitHub, so now we're in some grey circular dependency territory. I think I'd be ok with `./gradlew checkCompatibility -PrepositoryUrls=https://github.com/opensearch-project/job-scheduler.git,https://github.com/opensearch-project/common-utils.git,https://github.com/opensearch-project/asynchronous-search.git` though - can we get rid of the parts that create the circular parts and move those to a GHA?

**Comment by nknize** [comment]
`./gradlew checkCompatibility`

I actually love this. Just don't put it in `precommit` or `check`! Easy. It's optional. If I don't want to pull to mavenLocal I don't have to. And this is a tremendously useful tool to quickly check which "bundle" plugins are impacted from the change. This enables a developer (me, selfishly) to quickly and iteratively run the check if I'm breaking java namespaces (intentionally or unintentionally).

I actually view this much like the missingJavadoc checker. I use it quite often to ensure javadocs don't fail. We do enforce it on PR but I wonder how many actually run it locally? Just because not many use it (because they don't like writing javadocs or don't have a PR w/ any javadoc requirements) doesn't mean it isn't valuable to raise the bar here. 

I also like the ability to specify urls on the CLI or I can edit my own local .meta file. This is a nice flexible devops tool. 

The part I'm concerned about relying solely on GHA is the number of GHA calls we have. We have tightly coupled to GHA and I like having the CLI freedom to check things locally without having to open the PR first. It greatly cuts down on the amount of noise and minimizes the need for me to consistently rebase my PR (and trigger another GHA) to check for breaking changes.

**Review by nknize — APPROVED** [review]
I'm good merging this as is and we can iterate with follow on improvements. I'd like to start using this immediately as a mechanism for #8110 to help reduce breaking change fatigue. Any opposition to progress not perfection here @reta @andrross @dblock?

**Comment by andrross** [comment]
> Valid point but I doubt anyone will do that to be fair, in any case my main concern is polluting the OpenSearch build is not worth such feature (personal opinion).

@reta I agree in principle that OpenSearch should not have any knowledge about things that depend on it. However, the reality is that the bundle we ship today is pretty tightly coupled so I don't find this to be unreasonable. I would also doubt that anyone would use this if it is just an optional task that has to be manually invoked, but we've got one person here (@nknize) saying he wants to use this now. This is easily modified or reverted if it proves unnecessary or problematic in any way, so I'm fine with merging this.

File: buildSrc/build.gradle [hunk 1]
```diff
@@ -122,6 +122,8 @@ dependencies {
   api 'org.jruby.jcodings:jcodings:1.0.58'
   api 'org.jruby.joni:joni:2.1.48'
   api "com.fasterxml.jackson.core:jackson-databind:${props.getProperty('jackson_databind')}"
+  api "org.ajoberstar.grgit:grgit-core:5.2.0"
+
```
Comment (andrross) [inline buildSrc/build.gradle:126]: Nitpick: extra line

**Review by andrross — APPROVED** [review]

File: buildSrc/build.gradle (hunk 1 above)
Comment (gaiksaya) [reply buildSrc/build.gradle:126]: Fixed it in recent commit

**Comment by reta** [comment]
Thanks @andrross and @nknize , I have no objections, thanks @gaiksaya !

**Comment by nknize** [comment]
Thanks everyone! Thank you @gaiksaya for doing this!

**Comment by gaiksaya** [comment]
Thanks for review and merging this. While this works great in local (since we have or can set up SSH agent), I saw some issues while adding a GitHub Action workflow that will run the compatibility check on every PR and comment on the PR if any plugin is incompatible. 
The plugins list [here](https://raw.githubusercontent.com/opensearch-project/opensearch-plugins/main/.meta) contains SSH URLs. It becomes difficult to add such configuration as CI level. I tried adding a method to change those URLs from SSH to HTTPS. 

This is the sample run: https://github.com/gaiksaya/OpenSearch/actions/runs/5341380471/jobs/9682137303#step:4:1858

Let me know if anyone has objections proceeding with this.
cc: @bbarani 

Thanks!

**Comment by nknize** [comment]
It looks like the alerting failure is unrelated. I'm guessing this will just flag as "incompatible" if there is any kind of build failure regardless of whether it's related to the proposed change or not?

```
e: /tmp/groovy-generated-tmpdir-8549759542023875633/alerting/src/main/kotlin/org/opensearch/alerting/AlertService.kt: (522, 13): 'when' expression must be exhaustive, add necessary 'AUDIT' branch or 'else' branch instead

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':alerting:compileKotlin'.
> Compilation error. See log for more details
```

Also, it's hard to track which failure goes with which repo (e.g., the `cannot find symbol` for asynchronous-search).  It looks like the line `logger.info("Checking compatibility for: $repositoryUrls for $ref")` isn't showing. Does this need to switch to `logger.lifecycle("Checking compatibility for: $repositoryUrls for $ref")`?

**Comment by gaiksaya** [comment]
>It looks like the alerting failure is unrelated. I'm guessing this will just flag as "incompatible" if there is any kind of build failure regardless of whether it's related to the proposed change or not?

Yes! There will be false positives and some false negatives. If the plugin build itself is broken, it's going to fail no matter what. Plugin teams need to make sure their component's branch builds.
Some false positives are if you building with main branch as base but ran compatibility check for 2.x branch, its going to pass because publishToMavenLocal only published `3.0.0-snapshot`(which main points too) and plugin dependencies (2.0.0-snapshot) where grabbed from maven central

> Does this need to switch to logger.lifecycle("Checking compatibility for: $repositoryUrls for $ref")?

Yes will take this up in next PR. Thanks for pointing that out.

**Pull request merged**